lib/chef/provider/package.rb

```diff
@@ -509,7 +509,7 @@ def target_version_array
                   if version_requirement_satisfied?(current_version, new_version)
                     logger.trace("#{new_resource} #{package_name} #{current_version} satisifies #{new_version} requirement")
                     target_version_array.push(nil)
-                  elsif current_version && !allow_downgrade && version_compare(current_version, new_version) == 1
+                  elsif current_version && version_compare(current_version, new_version) == 1 && !allow_downgrade
```

Comment:
We have shifted !allow_downgrade check at the end since it always returns true if allow_downgrade is not provided or supported due to which the version_compare code is unreachable.

Reply:
That change needs to get reverted.  The existing behavior is correct.

This codeblock is all about preventing downgrades.  By default package providers now set this to true, which means they do not prevent downgrades, which is correct.  If a specific package provider wants to support this then it *must* implement the allow_downgrades property API in the specific provider -- and only if the user switches it to false manually will this code get triggered.

And then the choice to short circuit the version comparison is deliberate because the version comparison is not useful if allow_downgrade will always cause it to be skipped anyway.

Reply:
Got it thanks :+1: